Multi dof action extension

DLu opened the pull request:
Alternate approach for #16 

Adds `trajectory_msgs/MultiDOFJointTrajectory` to `FollowJointTrajectory.action` goal, as well as some related fields. 

Since this changes a core message, we should probably create a `galactic-devel` fork and merge there instead. However, adding the field shouldn't break too much since robots without a MultiDOF joint can just ignore the component.

bmagyar commented:
This is a fairly major change and I'd like to see some current motivation for this. 

Is it to add some cartesian constraints? Planning for a mobile manipulator including the base?

AndyZe commented:
Here's what @DLu is up to -- https://twitter.com/PickNikRobotics/status/1385667196999725056

AndyZe commented on control_msgs/action/FollowJointTrajectory.action:
```diff
@@ -1,5 +1,6 @@
 # The joint trajectory to follow
 trajectory_msgs/JointTrajectory trajectory
+trajectory_msgs/MultiDOFJointTrajectory multi_dof_trajectory
```
Could use a comment above each of these types to explain what they do and what the difference is. Something like --
```

# For simple revolute or prismatic joints only
trajectory_msgs/JointTrajectory trajectory
# For all joint types, including floating and planar
trajectory_msgs/MultiDOFJointTrajectory multi_dof_trajectory
```

DLu replied:
Added comments

AndyZe commented:
First thought is that the name is misleading. `MultiDOF` sounds like a normal 7-DOF arm with revolute joints. The name is also confusing because MoveIt supports a single joint having more than one degree of freedom (but it's rarely used).

Maybe call it `MixedDOFJointTrajectory`? Or `MixedTypeJointTrajectory`?

DLu commented:
> First thought is that the name is misleading. `MultiDOF` sounds like a normal 7-DOF arm with revolute joints. The name is also confusing because MoveIt supports a single joint having more than one degree of freedom, like a shoulder joint (but it's rarely used).
> 
> Maybe call it `MixedDOFJointTrajectory`? Or `MixedTypeJointTrajectory`?

This PR adds support for single joints that have more than one degree of freedom, i.e. planar/floating. It adds a `Trajectory` for `MultiDOFJoint`s, not a `JointTrajectory` with `MultiDOF`, if that makes sense. 

Either way, the naming reflects the existing names from [trajectory_msgs](http://docs.ros.org/en/api/trajectory_msgs/html/index-msg.html).

DLu commented:
Also, I've slightly modified the `MultiDOFJointTolerance` message to specify how to actually calculate the differences, based on the component field.

AndyZe commented on control_msgs/action/FollowJointTrajectory.action:
```diff
@@ -50,3 +54,7 @@ string[] joint_names
+
+string[] multi_dof_joint_names
+trajectory_msgs/MultiDOFJointTrajectoryPoint multi_dof_desired
+trajectory_msgs/MultiDOFJointTrajectoryPoint multi_dof_actual
```
Seems like there should be a --

`trajectory_msgs/JointTrajectoryPoint error`

to match L56

DLu replied:
Its redundant, but might as well.

AndyZe commented on control_msgs/msg/MultiDOFJointTolerance.msg:
```diff
@@ -0,0 +1,28 @@
+# For these components, the units are meters, meters/sec and meters/sec^2.
+# Z_AXIS is only valid with a floating joint, not planar.
+
+# If the component is ANGULAR the tolerance is measured in
```
ANGULAR doesn't exist. I think you mean ROTATION at L22?

DLu replied:
Good catch

AndyZe reviewed (approved):
I can't think of a better way to do this. Approve assuming you make the minor changes.

DLu commented:
What do you think @bmagyar ?

gavanderhoorn commented:
Is the rationale for extending the existing message the fact a multi-dof joint is still a "joint", as in contrast to Cartesian waypoints, trajectories (& associated goals) which do not really fit in at the joint level any more?

I'm asking because it doesn't necessarily seem proper to me to make `FollowJointTrajectory.action` a union capable of supporting all these different use-cases. Or perhaps in other words: does this semantically make sense? Are there any other future extensions expected?

bmagyar commented:
It definitely extends the scope of this message and opens a lot of possibilities for new use-cases.

The whole-body motion planning aspect is what was really convincing to me.

I have one ask though:
`multi_dof_path_tolerance` et al is a hazy name IMO. Can we come up with anything better than this to better describe it? We can't change the message type name but the local member name should be improved.

* `floating_joint_path_tolerance`?
* `transform_path_tolerance`?

DLu commented:
> Is the rationale for extending the existing message the fact a multi-dof joint is still a "joint", as in contrast to Cartesian (fi) waypoints, trajectories (& associated goals) which do not conceptually really fit in at the joint level any more?

The concept of "joint" is rather wide, but yes, the rationale is that planar and floating are the two multi-dof joints defined in the URDF spec, and are thus still "joints". Furthermore, MoveIt treats them as joints and can plan for them. 

> I'm asking because it doesn't necessarily seem proper to me to make `FollowJointTrajectory.action` a union capable of supporting all these different use-cases. 

Here's my thinking: Let's say we have two robots, R1 and R2, where R1 has MultiDOF joints and R2 does not. Both of them offer the `FollowJointTrajectory` action. 

Right now, R1 has no way to follow a trajectory that includes its MultiDOF joints. Extending the action definition like in this PR enables that capability. R2 does not need to change their implementation at all, because they can safely ignore the `multi_dof` field. Both robots can still use all nodes that employ the `FollowJointTrajectory` action. 

In the other case where we create a new action that includes the union (both MultiDOF and Regular Trajectory) (call it `FollowAllJointTrajectory.action)`, R1 needs to implement an additional action server, which looks a lot like the existing `FollowJointTrajectory` action server. But then all the other nodes that use `FollowJointTrajectory` may also need to be expanded to call either `FollowJointTrajectory` or `FollowAllJointTrajectory` in order to work with all the joints of R1. Furthermore, if a tool comes along that _only_ uses `FollowAllJointTrajectory` then R2 will have to write an additional action server as well in order to use that functionality. 


> Or perhaps in other words: does this semantically make sense? Are there any other future extensions expected? 

`trajectory_msgs` has been [functionally unchanged since 2013](https://github.com/ros/common_msgs/commits/noetic-devel/trajectory_msgs/msg) and its description states that they function as the [building blocks for most of control_msgs](https://github.com/ros/common_msgs/blob/20a833b56f9d7fd39655b8491a2ec1226d2639b3/trajectory_msgs/package.xml#L5-L7). As indicated by #16, this sort of functionality has been desired since at least 2015. 

> Should there be a separate `.action` for this, similar to how FZI created one for Cartesian use-cases?
I'm not sure what you're referring to here with FZI.

AndyZe commented:
The FZI reference is to this, I think:

https://discourse.ros.org/t/developing-a-cartesian-manipulator-interface/13872

The takeaway is, they made a new action server for Cartesian motions instead of modifying the existing one.

DLu commented:
> `multi_dof_path_tolerance` et al is a hazy name IMO. Can we come up with anything better than this to better describe it? We can't change the message type name but the local member name should be improved.

I made the message type name, so I went ahead and renamed it. Since the only difference between it and `JointTolerance.msg` is the component field, I renamed it `JointComponentTolerance.msg`

bmagyar commented:
@AndyZe thanks for that reference. 

The way I see it is that this would allow us to extend the FollowJointTrajectory action with non-revolute/prismatic/linear joints. The goal is not to make a shapeshifter follow trajectory action but to extend the current setup to support whole body trajectories where even in the case of bipeds where a pure joint trajectory ****could exist**** (however unrealistic), but the lower body is abstracted as a floating joint for practical reasons, similarly to mobile bases. This would play nicely with controller chaining once we have it in `ros2_control`.

DLu commented:
Thanks!

DLu commented:
(and thanks for the Galactic release)
